core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java

```diff
@@ -0,0 +1,405 @@
+          return o.getId();
+        }
+      };
+  private static final IndexDefinition<MasterInfo> HOSTNAME_INDEX =
+      new IndexDefinition<MasterInfo>(true) {
+        @Override
+        public Object getFieldValue(MasterInfo o) {
+          return o.getHostname();
+        }
+      };
+
+  /** Handle to the block master. */
+  private final BlockMaster mBlockMaster;
+
+  /** The clock to use for determining the time. */
+  private final Clock mClock = new SystemClock();
+
+  /** The connect address for the rpc server. */
+  private final InetSocketAddress mRpcConnectAddress
+      = NetworkAddressUtils.getConnectAddress(NetworkAddressUtils.ServiceType.MASTER_RPC);
+
+  /** is true if the master is in safe mode. */
+  private boolean mIsInSafeMode;
+
+  /** Keeps track of standby masters which are in communication with the leader master. */
+  private final IndexedSet<MasterInfo> mMasters =
+      new IndexedSet<>(ID_INDEX, HOSTNAME_INDEX);
+  /** Keeps track of standby masters which are no longer in communication with the leader master. */
+  private final IndexedSet<MasterInfo> mLostMasters =
+      new IndexedSet<>(ID_INDEX, HOSTNAME_INDEX);
+
+  /** The master configuration checker. */
+  private final ServerConfigurationChecker mMasterConfigChecker = new ServerConfigurationChecker();
+  /** The worker configuration checker. */
+  private final ServerConfigurationChecker mWorkerConfigChecker = new ServerConfigurationChecker();
+
+  /** The hostname of this master. */
+  private String mMasterHostname = Configuration.get(PropertyKey.MASTER_HOSTNAME);
+
+  /** The master ID for this master. */
+  private AtomicReference<Long> mMasterId = new AtomicReference<>(-1L);
+
+  /** Client for all meta master communication. */
+  private final MetaMasterMasterClient mMetaMasterClient
+      = new MetaMasterMasterClient(MasterClientConfig.defaults());
+
+  /** The start time for when the master started serving the RPC server. */
+  private long mStartTimeMs = -1;
+
+  /**
+   * Creates a new instance of {@link DefaultMetaMaster}.
+   *
+   * @param blockMaster a block master handle
+   * @param masterContext the context for Alluxio master
+   */
+  DefaultMetaMaster(BlockMaster blockMaster, MasterContext masterContext) {
+    this(blockMaster, masterContext, ExecutorServiceFactories
+        .fixedThreadPoolExecutorServiceFactory(Constants.META_MASTER_NAME, 2));
+  }
+
+  /**
+   * Creates a new instance of {@link DefaultMetaMaster}.
+   *
+   * @param blockMaster a block master handle
+   * @param masterContext the context for Alluxio master
+   * @param executorServiceFactory a factory for creating the executor service to use for running
+   *        maintenance threads
+   */
+  DefaultMetaMaster(BlockMaster blockMaster, MasterContext masterContext,
+      ExecutorServiceFactory executorServiceFactory) {
+    super(masterContext, new SystemClock(), executorServiceFactory);
+    mIsInSafeMode = masterContext.getmSafeModeManager().isInSafeMode();
```

Comment:
safe mode changes over time, this will only record whether we were in safe mode at the time that the constructor was called

Reply:
Save SafeModeManager in the DefaultMetaMaster and get isInSafeMode from SafeModeManager each time we need this info. Thanks